File: app/src/test/java/com/quran/labs/androidquran/model/bookmark/BookmarkModelTest.java
package com.quran.labs.androidquran.model.bookmark;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anySet;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.quran.data.model.bookmark.Tag;
import com.quran.labs.BaseTestExtension;
import com.quran.labs.androidquran.database.BookmarksDBAdapter;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.HashSet;

import io.reactivex.rxjava3.observers.TestObserver;

public class BookmarkModelTest {

  @Mock BookmarksDBAdapter bookmarksAdapter;
  @Mock RecentPageModel recentPageModel;
  private BookmarkModel model;

  @Before
  public void setupTest() {
    MockitoAnnotations.openMocks(BookmarkModelTest.this);
    model = new BookmarkModel(bookmarksAdapter, recentPageModel);
  }

  @Test
  public void testUpdateTag() {
    when(bookmarksAdapter.updateTag(anyLong(), anyString())).thenReturn(true);

    Tag tag = new Tag(1, "First Tag");
    TestObserver<Void> testObserver = new TestObserver<>();
    model.updateTag(tag)
        .subscribe(testObserver);
    BaseTestExtension.awaitTerminalEvent(testObserver);
    testObserver.assertNoErrors();
    testObserver.assertComplete();

    verify(bookmarksAdapter, times(1)).updateTag(tag.getId(), tag.getName());
  }

  @Test
  public void testUpdateBookmarkTags() {
    when(bookmarksAdapter.tagBookmarks(
        any(long[].class), anySet(), anyBoolean())).thenReturn(true);

    TestObserver<Boolean> testObserver = new TestObserver<>();
    model.updateBookmarkTags(new long[] { }, new HashSet<>(), false)
        .subscribe(testObserver);
    BaseTestExtension.awaitTerminalEvent(testObserver);
    testObserver.assertNoErrors();
    testObserver.assertComplete();
  }
}

File: app/src/test/kotlin/com/quran/labs/androidquran/model/bookmark/BookmarkModelTest.kt
package com.quran.labs.androidquran.model.bookmark

import app.cash.sqldelight.adapter.primitive.IntColumnAdapter
import app.cash.sqldelight.driver.jdbc.sqlite.JdbcSqliteDriver
import com.google.common.truth.Truth.assertThat
import com.quran.data.model.bookmark.Tag
import com.quran.labs.awaitTerminalEvent
import com.quran.labs.androidquran.BookmarksDatabase
import com.quran.labs.androidquran.database.BookmarksDBAdapter
import com.quran.mobile.bookmark.Bookmarks
import com.quran.mobile.bookmark.Last_pages
import com.quran.labs.androidquran.base.TestApplication
import com.quran.labs.test.RxSchedulerRule
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Tests for BookmarkModel using in-memory SQLite database.
 * Migrated from Mockito to real database implementation.
 *
 * Uses RobolectricTestRunner so that the SQLite JDBC driver loads in the same sandbox
 * classloader as other Robolectric tests, preventing DriverManager classloader conflicts
 * when the full test suite runs.
 */
@Config(application = TestApplication::class, sdk = [33])
@RunWith(RobolectricTestRunner::class)
class BookmarkModelTest {

  @get:Rule
  val rxRule = RxSchedulerRule()

  private lateinit var database: BookmarksDatabase
  private lateinit var bookmarksAdapter: BookmarksDBAdapter
  private lateinit var recentPageModel: RecentPageModel
  private lateinit var model: BookmarkModel

  @Before
  fun setupTest() {
    val driver = JdbcSqliteDriver(JdbcSqliteDriver.IN_MEMORY)
    BookmarksDatabase.Schema.create(driver)
    database = BookmarksDatabase(
      driver,
      Bookmarks.Adapter(IntColumnAdapter, IntColumnAdapter, IntColumnAdapter),
      Last_pages.Adapter(IntColumnAdapter)
    )
    bookmarksAdapter = BookmarksDBAdapter(database)
    recentPageModel = RecentPageModel(bookmarksAdapter)
    model = BookmarkModel(bookmarksAdapter, recentPageModel)
  }

  @Test
  fun testUpdateTag() {
    // Setup: Create a tag first
    val tagId = bookmarksAdapter.addTag("Original Tag")
    val tag = Tag(tagId, "Updated Tag")

    // Execute: Update the tag
    val testObserver = model.updateTag(tag).test()
    testObserver.awaitTerminalEvent()

    // Verify: No errors and operation completed
    testObserver.assertNoErrors()
    testObserver.assertComplete()

    // Verify: Tag was actually updated in database
    val allTags = bookmarksAdapter.getTags()
    val updatedTag = allTags.find { it.id == tagId }
    assertThat(updatedTag).isNotNull()
    assertThat(updatedTag!!.name).isEqualTo("Updated Tag")
  }

  @Test
  fun testUpdateBookmarkTags() {
    // Setup: Create a bookmark and a tag
    val bookmarkId = bookmarksAdapter.addBookmark(null, null, 1)
    val tagId = bookmarksAdapter.addTag("Test Tag")
    val tags = setOf(tagId)

    // Execute: Tag the bookmark
    val testObserver = model.updateBookmarkTags(longArrayOf(bookmarkId), tags, false).test()
    testObserver.awaitTerminalEvent()

    // Verify: No errors and operation completed successfully
    testObserver.assertNoErrors()
    testObserver.assertComplete()
    testObserver.assertValue(true)

    // Verify: Bookmark is actually tagged in database
    val bookmarkTags = bookmarksAdapter.getBookmarkTagIds(bookmarkId)
    assertThat(bookmarkTags).contains(tagId)
  }
}
